v2 master -- updates to basicSampling, simpleFunP, some tests, and some examples

eecsu opened the pull request:
Validation, linear, and nonlinear examples updated.

Added regular sampling to basicSampling. Still need to add tests, but using these without problems in the examples.

Added different types of functions to simpleFunP and added tests for all of these.

lcgraham commented on bet/calculateP/simpleFunP.py:
```diff
@@ -17,25 +17,26 @@ class wrong_argument_type(Exception):
-    the size of the generalized rectangle by scaling the circumscribing 
-    generalized rectangle of :math:`\mathcal{D}`.
-    The simple function approximation is then defined by determining M 
+    The support of this density is defined by rect_size, which determines
```
rect _size should be ``rect_size`` for stylistic reasons

eecsu replied:
Where are you seeing "rect _size"? I only see "rect_size".

lcgraham replied:
It's adding the double single quotes since it's an variable/attribute name so that it will be formatted as code

> On May 23, 2016, at 12:28 PM, Troy Butler <notifications@github.com> wrote:
> 
> In bet/calculateP/simpleFunP.py:
> 
> >      r"""
> >      Creates a simple function approximation of :math:`\rho_{\mathcal{D}}`
> >      where :math:`\rho_{\mathcal{D}}` is a uniform probability density on
> >      a generalized rectangle centered at Q_ref.
> > -    The support of this density is defined by bin_ratio, which determines
> > -    the size of the generalized rectangle by scaling the circumscribing 
> > -    generalized rectangle of :math:`\mathcal{D}`.
> > -    The simple function approximation is then defined by determining M 
> > +    The support of this density is defined by rect_size, which determines
> Where are you seeing "rect _size"? I only see "rect_size".
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
>

eecsu replied:
Okay. I will make a lot of those changes then. It was pretty inconsistently applied in a lot of commenting that I had not written but used to copy/paste. Same should be true for Q-ref vs ``Q_ref``, right?

lcgraham replied:
Yeah, I should really go through all of the docstrings and make sure that
this is consistent. I think it makes the documentation more readable. If
you have time to do it when you alter an existing method I would recommend
it. I will look through the rest of the code later to check for
consistency. This is something that is on my todo list for rebuilding or
updating the documentation when we merge v2 into master.
On Mon, May 23, 2016 at 1:22 PM Troy Butler <notifications@github.com>
wrote:

> In bet/calculateP/simpleFunP.py
> <https://github.com/UT-CHG/BET/pull/195#discussion_r64259970>:
>
> >      r"""
> >      Creates a simple function approximation of :math:`\rho_{\mathcal{D}}`
> >      where :math:`\rho_{\mathcal{D}}` is a uniform probability density on
> >      a generalized rectangle centered at Q_ref.
> > -    The support of this density is defined by bin_ratio, which determines
> > -    the size of the generalized rectangle by scaling the circumscribing
> > -    generalized rectangle of :math:`\mathcal{D}`.
> > -    The simple function approximation is then defined by determining M
> > +    The support of this density is defined by rect_size, which determines
>
> Okay. I will make a lot of those changes then. It was pretty
> inconsistently applied in a lot of commenting that I had not written but
> used to copy/paste. Same should be true for Q-ref vs Q_ref, right?
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/UT-CHG/BET/pull/195/files/3168f991c93453e42a622b2180bd5880118f59a9#r64259970>
>
-- 

Lindley Graham, Ph.D. 2015

eecsu replied:
Well I will try to remember to do this. Some finer details like this are pretty low on my priority list as we move towards the functional v2_master that will be the Beta release. Once that is done, then the grunt work of fixing up all the documentation that we either forgot to do or skipped during the v2_master overhaul will get addressed before the official v2 release. At least this was what I had in mind.

lcgraham replied:
Yup my thoughts exactly.

> On May 23, 2016, at 3:58 PM, Troy Butler <notifications@github.com> wrote:
> 
> In bet/calculateP/simpleFunP.py:
> 
> >      r"""
> >      Creates a simple function approximation of :math:`\rho_{\mathcal{D}}`
> >      where :math:`\rho_{\mathcal{D}}` is a uniform probability density on
> >      a generalized rectangle centered at Q_ref.
> > -    The support of this density is defined by bin_ratio, which determines
> > -    the size of the generalized rectangle by scaling the circumscribing 
> > -    generalized rectangle of :math:`\mathcal{D}`.
> > -    The simple function approximation is then defined by determining M 
> > +    The support of this density is defined by rect_size, which determines
> Well I will try to remember to do this. Some finer details like this are pretty low on my priority list as we move towards the functional v2_master that will be the Beta release. Once that is done, then the grunt work of fixing up all the documentation that we either forgot to do or skipped during the v2_master overhaul will get addressed before the official v2 release. At least this was what I had in mind.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
>

lcgraham commented on bet/sampling/basicSampling.py:
```diff
@@ -212,8 +214,106 @@ def random_sample_set_dimension(self, sample_type, input_dim,
+        :type num_samples_per_dim: :class: `~numpy.ndarray` of dimension
+            (input_sample_set._dim,)
+
+        :rtype: :class:`~bet.sample.discretization`
```
this actually returns ``bet.sample.sample_set``

lcgraham commented on bet/sampling/basicSampling.py:
```diff
@@ -212,8 +214,106 @@ def random_sample_set_dimension(self, sample_type, input_dim,
+        :type num_samples_per_dim: :class: `~numpy.ndarray` of dimension
+            (input_sample_set._dim,)
+
+        :rtype: :class:`~bet.sample.discretization`
```
this actually returns ``bet.sample.sample_set``

lcgraham commented on bet/sampling/basicSampling.py:
```diff
@@ -212,8 +214,106 @@ def random_sample_set_dimension(self, sample_type, input_dim,
+            (input_sample_set._dim,)
+
+        :rtype: :class:`~bet.sample.discretization`
+        :returns: :class:`~bet.sample.discretization` object which contains
```
this actually returns ``bet.sample.sample_set``

eecsu replied:
Will fix. Copy/paste typos.

lcgraham commented on bet/sampling/basicSampling.py:
```diff
@@ -212,8 +214,106 @@ def random_sample_set_dimension(self, sample_type, input_dim,
+                input_domain[i,0], input_domain[i,1],
+                num_samples_per_dim[i]+2))[1:num_samples_per_dim[i]+1]
+
+        arrays_samples_dimension = np.meshgrid(
```
This is fantastic! We should use this syntax to rewrite ``bet.util.meshgrid``.

eecsu replied:
I done did good! I played around with this in ipython a bit before I had it as compact/smooth as possible. Glad you like it.

lcgraham commented on bet/calculateP/simpleFunP.py:
```diff
@@ -75,39 +76,41 @@ def unif_unif(data_set, Q_ref, M=50, bin_ratio=0.2, num_d_emulate=1E6):
+    if not isinstance(rect_size, collections.Iterable):
+        rect_size = rect_size * np.ones((dim,))
+    if np.any(np.less(rect_size, 0)):
+        print 'Warning: rect_size must be greater than 0'
```
These print lines should be replaced with ``logging.warn``, you will need to add an ``import logging`` at the top of the module

eecsu replied:
Yeah, I saw a few places you used it that caused some merge conflicts with me previously that I just didn't want to deal with right now since I had not used the logging before. I will incorporate those. Should this actually be an error though? I think the check should actually be ``np.any(np.less_equal(rect_size, 0))`` and this should be an error because it makes no sense if the ``rect_size`` is less or equal to 0.

lcgraham replied:
Yes, this should be an exception.

lcgraham commented:
I know you still need to add tests for the new methods in ``basicSampling``, but since we are now doing continuous integration on this branch I would like those tests to be added before merging this pull request.

Tests are also missing for ``uniform_partition_normal_distribution`` and ``user_discretization_user_distribution``.

eecsu commented:
I will work on those tests now while my son is napping. Hopefully can knock them out quickly.

lcgraham commented on bet/sampling/basicSampling.py:
```diff
@@ -227,17 +227,17 @@ def regular_sample_set(self, input_sample_set, num_samples_per_dim=1):
         if not isinstance(num_samples_per_dim, collections.Iterable):
             num_samples_per_dim = num_samples_per_dim * np.ones((dim,))
-        if np.any(np.less(num_samples_per_dim, 0)):
+        if np.any(np.less_equal(num_samples_per_dim, 0)):
```
Throw an exception here. (No need to use logging since it's an exception)

eecsu commented:
I think I have addressed all the issues you raised. No logging but lots of exceptions raised if input arguments are illogical (like empty domains being specified due to a rect_size <=0). We had lots of messages print out before, but these are just flat out user errors.
